dask/array/core.py

```diff
@@ -2833,20 +2833,21 @@ def where(condition, x=None, y=None):
     if x is None or y is None:
         raise TypeError(where_error_message)
 
-    x = asarray(x)
-    y = asarray(y)
-
-    shape = broadcast_shapes(x.shape, y.shape)
-    dtype = np.promote_types(x.dtype, y.dtype)
+    if np.isscalar(condition):
+        # Match the dtype inference logic from elemwise
+        vals = [np.empty((1,) * a.ndim, dtype=a.dtype)
+                if not is_scalar_for_elemwise(a) else a
+                for a in (x, y)]
+        dtype = apply_infer_dtype(np.where, [condition] + vals, {}, 'where', suggest_dtype=False)
```

Comment:
Using `promote_types` gives the wrong type if one of the inputs (x or y) is a scalar, because the numpy (and `elemwise`) promotion rules are different for scalars. One way to simplify the code would be to take this hoist this bit of logic out of `elemwise` and into its own function, so that this function can just call on it.

As for unconditional `broadcast_shapes`, try running the following on 0.15.1 and looking at the `mydask.png` that it produces: there are a whole bunch of nodes for rechunking of the broadcast of `b` - it's creating extra nodes for every chunk in `a`. With the new code, those nodes go away.